docs/entrypoints.md

```diff
@@ -0,0 +1,70 @@
+# Entrypoints
+
+This document describes the various ways that weave as a binary gets launched, and where and how it processes options.
+
+The goal is to enable maintainers to modify functionality without having to reverse engineer the code to find all of the places you can start a weave binary and what it depends upon.
+
+## Components
+### weaver
+
+`weaver` is the primary binary: it sets up the weave bridge to which containers are attached; manages IP address allocation, serves DNS requests and does many more things.
+
+Since version 2.0, `weaver` also bundles the Docker Plugin (legacy version) and Docker API Proxy which used to be shipped as separate containers.
+
+### weaveutil
+weaveutil is a binary which provides a number of functions for managing an existing weave network. It gets information about the network, attaches and detaches containers, etc.
+
+For the majority of interactions with an existing weave network, you will launch `weaveutil` in some manner or another.
+
+Almost all options can be passed as `--option` to `weaveutil`. These options are created in `prog/weaveutil/main.go` with a table mapping each command to a dedicated golang function.
+
+In addition to operating in normal mode, `weaveutil` has several additional operating modes. Before processing any commands, `weaveutil` checks the filename with which it was called.
+
+* If it was `weave-ipam` it delegates responsibility to the `cniIPAM()` function in `prog/weaveutil/cni.go`, which, in turn, calls the standard CNI plugin function `cni.PluginMain()`, passing it weave's IPAM implementation from `plugin/ipam`.
+* If it was `weave-net` it delegates responsibility to the `cniNet()` function in `prog/weaveutil/cni.go`, which, in turn, calls the standard CNI plugin function `cni.PluginMain()`, passing it weave's net plugin implementation from `plugin/net`.
+
+Finally, `weaveutil` is called by `weaver` when it needs to operatein a different network namespace.
+Go programs cannot safely switch namespaces; see this [issue](http://github.com/vishvananda/netns/issues/17).
+The call chain is `weaver`->`nsenter`->`weaveutil`.
+
+### weave
+Wrapping `weaver` and `weaveutil` is `weave`, a `sh` script that provides help information and calls `weaveutil` as relevant.
+
```

Comment:
Suggestion: Add a comment about the history behind `weave` and `weaveutil`. Maybe something like the below?

> In versions of Weave Net prior to 2.0, `weave` used to have many utility functions implemented as simple shell functions, sometimes delegating to other Linux command line tools. 
> Current direction is to move some of this logic from `sh` to Go, for some or all of the below reasons: 
> - so it can re-use 3rd party libraries available in Go (e.g. to interact with Docker, Kubernetes, etc.),
> - so it can be re-used more conveniently in the various entry points, e.g.: `weave` and `launch.sh` (see below),
> - for better robustness and increased control over the flow of the program,
> - for better performance,
> - for better synchronisation and concurrency,
> - so it can be refactored and called directly from other components implemented in Go at a later point.
> 
> As a result, some of the utility functions originally implemented as `sh` functions in `weave` were moved to `weaveutil`.

Do you think this could be useful to new Weave Net developers @bboreham, @deitch?